fix comment parsing

Philipp-Binder opened the pull request:
Hi together,

I investigated how to replace the comment-detection via split with direct parser-functionality.

Doing this I think I replaced some bugs too. I've tried to document this in the first commit by adding more ParserTests showing the buggy behaviour and adding todos for recommended "fixed" tests.

Would be great to hear your opinion regarding the introduced tests.

Philipp-Binder committed:
add new tests to document current behavior; include todos for tests which I guess is buggy behavior;

Philipp-Binder committed:
parse comments by parser instead of splitting; fix introduced todos (bug-definitions) in ParserTests; trim remains, since UnquotedValue should always be trimmed;

rogusdev commented:
I will take a look at this at some point soon. Thank you for investigating and presenting a potential solution, with more tests!

Philipp-Binder committed:
fixed a bug with preceeding whitespaces or tabs before unmatched singleQuotes, unmatched doubleQuotes

Philipp-Binder commented on src/DotNetEnv/Parsers.cs:
```diff
@@ -173,17 +173,23 @@ private static string ToUtf32Char (string hex)
+                    .Select(x => new ValueActual(x))
+                )
+                .Many()
+                .Except(Parse.Regex("[ \t]*['\"#]"))
```
Except after Many resolves to an initial check whether we directly begin with an unmatched quote or a comment.

The following Or assures, that we never fail in here and return empty values if nothing matches. Invalid contents are handled by Assignment then.

rogusdev commented on src/DotNetEnv/Parsers.cs:
```diff
@@ -173,17 +173,23 @@ private static string ToUtf32Char (string hex)
                 .Or(EscapedChar);
 
+        internal static readonly Parser<IEnumerable<char>> InlineCommentBeginOrControl =
+            Parse.Regex("[ \t]#").Or(Parse.Char(c => char.IsControl(c) && c != '\t', "controls except tab").Once());
```
So, one of the goals I had with using the `Parser` approach was actually to avoid `Regex`. I am not 100% certain it can be avoided here, as it has been a very long time since I really thought about this code, so understanding your changes is a bit tricky for me at the moment, but my gut reaction is that I don't want to introduce regexes for handling comments differently than they are now, ideally.

rogusdev replied:
Looking closer at the changes, I see that I already had the regex and you were just carrying it over, so that makes total sense. I still would love a solution that was cleaner, ideally, but my bigger concern at this point is the seemingly new behavior below, of silently discarding invalid lines.

Philipp-Binder replied:
I think you can do something like `Parse.String(" #").Or(Parse.String("\t#")).Or(...)`. (at least the tests still succeed, so it should be correct I guess)
In my opinion such a short Regex is cleaner, but that's personal preference too.

About other concerns I will comment at the appropriate comments ;)

rogusdev commented on test/DotNetEnv.Tests/ParserTests.cs:
```diff
@@ -238,8 +249,22 @@ public void ParseUnquotedValue ()
 
-            Assert.Throws<ParseException>(() => Parsers.UnquotedValue.End().Parse("0\n1"));
-            Assert.Throws<ParseException>(() => Parsers.UnquotedValue.End().Parse("'"));
+            Assert.Equal("0", Parsers.UnquotedValue.Parse("0\n1").Value);   // value ends on linebreak
```
So, are these desirable changes? I explicitly want the system to throw up on bad input. Are there other examples in these tests already that are like these cases, where bad input should be silently processed and discarded?

Philipp-Binder replied:
I will have a closer look here.
I don't think I silently process and discard values, but the tests did not succedd in this form due to the changes.
I just forgot about these tests, I should have replaced them approprietly.

Philipp-Binder replied:
I will just add the both removed tests again, they still work.
I think I just forgot about them, I removed them at a point where I did not fully understand how these parsers work, and added my new tests afterwards.

If you force UnquotedValue to parse to end it fails of course, because the linebreak and text in the new line are not expected.

Explanation below, I will update the test to explain the intention better.

First additional test (value ends on linebreak):
If you just try to get the valid value before the linebreak (so not End, just parse) it should of course ignore the next line.
The next line is an invalid line then of course, but that's handled by analysis of the next line, the unquoted value of our tested line ist still perfectly valid.

Second additional test (unmatched singleQuote):
With just parsing an unquotedValue should not allow a leading singleQuote, so it stops directly and returns empty string.
The unmatched doubleQuote will then fail with the next parsers, because after having a value it's not allowed to have a singlequote on the same line.
For these cases we can also argue to throw, that could remove the requirement to have quoted parsers before unquoted in ValueParser I guess.

rogusdev replied:
I double checked and this test passes already in `master` because without `End()` it just means "this is how much it consumed" and the newline is not consumed. Maybe that's a useful test to add -- I added it to my PR out of respect for your enthusiasm for it -- but it is not testing new behavior.

Philipp-Binder replied:
Here the following two comments apply: https://github.com/tonerdo/dotnet-env/pull/87#discussion_r1437763489 and https://github.com/tonerdo/dotnet-env/pull/87#discussion_r1437790418

As I wrote there too, this test shows the "real" behavior better than the "End"-Tests, but after all I guess we do not need to be on the same page about everything ;)

rogusdev commented on test/DotNetEnv.Tests/ParserTests.cs:
```diff
@@ -340,6 +366,15 @@ public void ParseAssignment ()
+            testParse("EV_DNE", "test", "EV_DNE= test #basic comment");
+            testParse("EV_DNE", "test", "EV_DNE= test  #a'bc allow singleQuotes in comment");
+            testParse("EV_DNE", "test", "EV_DNE= test  #a\"bc allow doubleQuotes in comment");
+            testParse("EV_DNE", "test", "EV_DNE= test  #a$bc allow dollarSign in comment");
```
This one already passes, but seems like a good additional test

Philipp-Binder replied:
Yes, needed some tests youbt these values if they still succeed with the new changes.

rogusdev commented on test/DotNetEnv.Tests/ParserTests.cs:
```diff
@@ -340,6 +366,15 @@ public void ParseAssignment ()
 
+            testParse("EV_DNE", "", "EV_DNE=#no value just comment");
+            testParse("EV_DNE", "", "EV_DNE= #no value just comment");
+            testParse("EV_DNE", "test", "EV_DNE= test #basic comment");
```
This one already passes, but seems like a good additional test

rogusdev commented on test/DotNetEnv.Tests/ParserTests.cs:
```diff
@@ -340,6 +366,15 @@ public void ParseAssignment ()
                 Assert.Equal(value, kvp.Value);
             };
 
+            testParse("EV_DNE", "", "EV_DNE=#no value just comment");
```
These new tests all seem excellent, and (other than the 2 that already pass) seem to directly test the desired new behavior.

Philipp-Binder replied:
Just have in mind, this is a functional change.
In my opinion this is new desired behavior, but worth a not that this comment was interpreted as value before the change.

rogusdev commented on test/DotNetEnv.Tests/ParserTests.cs:
```diff
@@ -309,8 +334,9 @@ public void ParseValue ()
 
-            Assert.Throws<ParseException>(() => Parsers.Value.End().Parse("0\n1"));
-            Assert.Throws<ParseException>(() => Parsers.Value.End().Parse("'"));
+            Assert.Equal("0", Parsers.Value.Parse("0\n1").Value);   // value ends on linebreak
```
Again, I am very uncomfortable with introducing these behaviors.

Philipp-Binder replied:
And again I will definitely have a closer look and add tests and/or explain better ;)

Philipp-Binder replied:
Added the test in here too.

I will try to rewrite Unquoted-Matcher to throw on leading quotes, thinking about it this will be much cleaner (if you have to explain your tests so deeply may shows there is some problem in the code I guess :D )

rogusdev replied:
Part of the problem with these tests (now that I have refreshed myself on all this code) is that they should never happen. If the larger context sent this kind of text to be parsed by an unquoted value (i.e. with a newline in the middle), something has gone very wrong in that larger context, which is the Assignment parser.

Which is to further say that if this did happen, it should throw an error, ideally, because things are definitely broken somewhere. But it also should not happen, and I am disinclined to add tests for this sort of case, and definitely disinclined to change any code to support this case.

Philipp-Binder replied:
I adapted some of your code about the comment detection, because Sprache makes it really hard to have a working positive look-ahead. (which in fact is done by the combination of first parsing whitespaces and then throwing if we have a `hash`-character)

And there are quite some tests which enforce throwing exceptions at all the cases where we do not have a valid UnquotedValue-Parsing, even more than before, because before we just relied on the Assignment-Parser to consume all whitespaces before ValueParser, but now it is enforced by tests.

rogusdev commented on test/DotNetEnv.Tests/ParserTests.cs:
```diff
@@ -238,8 +249,22 @@ public void ParseUnquotedValue ()
-            Assert.Throws<ParseException>(() => Parsers.UnquotedValue.End().Parse("0\n1"));
-            Assert.Throws<ParseException>(() => Parsers.UnquotedValue.End().Parse("'"));
+            Assert.Equal("0", Parsers.UnquotedValue.Parse("0\n1").Value);   // value ends on linebreak
+            Assert.Equal("", Parsers.UnquotedValue.Parse("'").Value);  // unmatched singleQuote
```
I am not sure that this is how this should work -- it seems like this should be an error to me. Do you feel differently / do you see other examples in here that you think makes this consistent with existing behaviors?

To be clear: I don't remember all the examples in here very well at this point, so it may very well be the case that this is entirely consistent and I am thinking about things all wrong. :)

Philipp-Binder replied:
Due to rebasing I'm not completely sure if that's just a dupe, but I will look through the changes if there are more of these tests and reintroduce them appropriatly.

Philipp-Binder replied:
Checked it, there are no other occurences.

rogusdev commented on test/DotNetEnv.Tests/ParserTests.cs:
```diff
@@ -238,8 +249,22 @@ public void ParseUnquotedValue ()
+            Assert.Equal("", Parsers.UnquotedValue.Parse("   \"unmatched doubleQuote with preceding whitespace").Value);
+            Assert.Equal("", Parsers.UnquotedValue.Parse("\t\t\"unmatched doubleQuote with preceding tabs").Value);
+
+            Assert.Equal("", Parsers.UnquotedValue.Parse("#").Value); // no value, empty comment
```
This and the following examples about empty values followed by comments are all certainly excellent additions.

That said, how different are each of these from those in the `ParseEmpty` tests? Perhaps these should live there anyway?

rogusdev replied:
In fact, if you rewrite these first two in the style of `ParseEmpty` they pass already:
```
            Assert.Equal(kvp, Parsers.Empty.End().Parse("#"));
            Assert.Equal(kvp, Parsers.Empty.End().Parse("#commentOnly"));
```

And the following 3 tests already pass. (And have very similar matching examples in `ParseEmpty`)

Now I'm trying to remember why the `ParseEmpty` tests verify that the value is `null`, rather than `""`

rogusdev replied:
It is especially interesting to me that the last test case in `ParseEmpty` is effectively already this, but that test passes, whereas this one does not. I think that is the thread most worth pulling on. Because ideally, imo, any test in `ParseEmpty` returns `null`, should return `""` in `ParseUnquotedValue` and then a lot of these tests would just pass immediately...

Philipp-Binder replied:
The `UnquotedValue`-Parser should parse something after an `EqualSign`, so only "Values" it is used within `Assignment`-Parser inside the `Value`-Parser).

The `Empty`-Parser should Parse a whole Line, an it is used only inside the `ParseDotenvFile`-Method directly. And it has to be null to be able to exclude these entries afterwards.

Back to the Tests:
I should definitely have a closer look here too I guess, I get some feeling that there is something I just missed while making the changes which has to be revealed (connected to the removed test in the other comments).
I will update the PR in a few days, giving more explanation and I guess some code changes.

rogusdev commented:
Heads up that in investigating this, I did some trivial clean up that will require rebasing: https://github.com/tonerdo/dotnet-env/pull/85

rogusdev commented:
I had a bit of inspiration and put together: https://github.com/tonerdo/dotnet-env/pull/87 -- which does everything I was aiming for, to my perception: no regex splitting required to get comments on unquoted strings.

But it highlights that we now have one ambiguous case:
```
            testParse("EV_DNE", "", "EV_DNE=#no value just comment");
```

I have left this failing in this branch, because we need to talk about it. (I believe you touched on this with https://github.com/tonerdo/dotnet-env/pull/82#discussion_r1424412109 )

Consider 2 similar examples back to back (also in the PR)
```
            testParse("EV_DNE", "http://www.google.com/#anchor", "EV_DNE=http://www.google.com/#anchor");
            testParse("EV_DNE", "#anchor", "EV_DNE=#anchor");
```

Now, of these two, the first one should be unambiguous: we want a URL that happens to include a # char, but is just part of the unquoted string -- this is exactly a planned for use case.

The second of these two gets a bit fuzzier, but in the context of the first one, would seem reasonable that we want the ability to **start a value** with a # because sometimes that is a valid value.

However... obviously this means that if someone wanted to do:
```
            testParse("EV_DNE", "", "EV_DNE=#http://www.google.com/#anchor");
```

I.e. they wanted to comment out the entire value for EV_DNE in the .env file, *but not the entire line itself* then they could not do that, if we allowed values to start with a # (like the `#anchor` example above).

So: do we want values to be allowed to start with a #? And if you want to comment out a value, you just comment out the entire line, and put an extra line that sets `ENVVAR=` (i.e. empty) or `ENVVAR= #comment` which for sure works?

Or: do we want #s to only be valid in the middle of a string and not at the beginning?

And does this cover all the cases you were thinking of? I'm not quite sure yet.

rogusdev commented on test/DotNetEnv.Tests/ParserTests.cs:
```diff
@@ -340,6 +366,15 @@ public void ParseAssignment ()
+            testParse("EV_DNE", "test", "EV_DNE= test  #a'bc allow singleQuotes in comment");
+            testParse("EV_DNE", "test", "EV_DNE= test  #a\"bc allow doubleQuotes in comment");
+            testParse("EV_DNE", "test", "EV_DNE= test  #a$bc allow dollarSign in comment");
+            testParse("EV_DNE", "a'b''c", "EV_DNE=a'b''c #allow inline singleQuotes in unquoted values");
```
I'm not sure if these 2 are desirable, now that I reflect on it. Is there a reason you think these should be allowed?

rogusdev replied:
It would be very easy to add support to my changes for this feature, I just question its necessity.

rogusdev replied:
Of particular note, what would you want for:
```
            testParse("EV_DNE", "a 'b' c", "EV_DNE=a 'b' c #allow inline singleQuotes in unquoted values");
            testParse("EV_DNE", "a \"b\" c", "EV_DNE=a \"b\" c #allow inline doubleQuotes in unquoted values");
```

Should those pass?

If we are ok with these 2 failing, then this covers the examples you gave:
```
        internal static readonly Parser<ValueCalculator> UnquotedValue =
            InterpolatedValue.Or(
                InlineWhitespaceChars.Many().Then(w =>
                    NotControlNorWhitespaceChar("#'\"$").Once().Then(
                        c => NotControlNorWhitespaceChar("$").Many().Select(
                            cs => w.Concat(c).Concat(cs)
                        )
                    )
                ).Text()
                .Select(s => new ValueActual(s))
            ).Many().Select(vs => new ValueCalculator(vs));
```

However, I do wonder if we should support one, or both, or neither of these sets of examples -- with unquoted values. All of these already work with quotes:
```
            testParse("EV_DNE", "a'b''c", "EV_DNE=\"a'b''c\" #this already works");
            testParse("EV_DNE", "a\"b\"\"c", "EV_DNE='a\"b\"\"c' #this already works");
            testParse("EV_DNE", "a 'b' c", "EV_DNE=\"a 'b' c\" #this already works");
            testParse("EV_DNE", "a \"b\" c", "EV_DNE='a \"b\" c' #this already works");
```

Philipp-Binder replied:
Quotes within unquoted values is "solved": we will not support that. (see discussion here and in #87)

rogusdev commented:
OK, I think I talked myself into changing the behavior to make `EV_DNE=#no value just comment` truly just a comment, because in the current version of things, this fails:
```
            testParse("EV_DNE", "", "EV_DNE= #no value just comment");
```

Notice the space in front of the `#` -- I would **definitely** want this to be a comment, no question, and therefore this should pass, and that suggests to me that `EV_DNE=#comment` should also pass and with no value.

I have updated my PR accordingly

Philipp-Binder committed:
fix ParseInlineCommentBeginOrControl (xunit or framework itself seems to be more strict about equality of strings and char-arrays now)

Philipp-Binder committed:
add throwing "End()" tests in ParseUnquotedValue-Test again

Philipp-Binder committed:
add throwing "End()" tests in ParseValue-Test again

Philipp-Binder committed:
add explicit tests for singleQuotes inside doubleQuotes and vice versa

Philipp-Binder committed:
use Not() to return empty-values only if there is a comment-only value, but not for invalid values

rogusdev commented:
@Philipp-Binder before you go further on this PR -- tho I hope you are having a fun! :) -- I would like you to review my PR as an alternative proposal that I believe is a simpler and cleaner solution: #87 and give me any feedback on that, if you think there is anything missing there such that your approach is preferred, please.

What I mean is: do you think there is anything missing in my PR that you think your PR handles? I don't believe we need both.

rogusdev commented on src/DotNetEnv/Parsers.cs:
```diff
@@ -173,17 +173,23 @@ private static string ToUtf32Char (string hex)
-                .Select(strs => new ValueActual(strs))
-            ).Many().Select(vs => new ValueCalculator(vs).Split("[ \t]#").Trim());
+            InterpolatedValue
+                .Or(Parse.CharExcept('$').Except(InlineCommentBeginOrControl)
```
This is the bit that I feel makes your current approach more complicated than mine in #87 -- you attempted to retain too much of the old (undesired) behavior of the previous implementation, which is that the unquoted value continues to parse comments itself, and then discard them, rather than leaving that responsibility to the Assignment parser which discards comments for single + double quoted comments already.

My PR changes it so that the unquoted value parser *does not read comments at all* so the Assignment parser continues to do it's job the way it's supposed to, and discards comments there, consistent with the other kinds of values.

Your PR continues to parse comments as a part of unquoted value, which is, at the core, the behavior that the FIXME comment was aiming at stopping (as identified as being about the regex / splitting and trimming).

Philipp-Binder replied:
Changed an adapted some of your code, I'm closing this, if there are additional things just reopen or add new discussion ;)

Philipp-Binder commented:
@rogusdev 
> What I mean is: do you think there is anything missing in my PR that you think your PR handles? I don't believe we need both.

I think the duscussion moves us in the right direction, but there are still some problems in both solutions.
In the end we will need only one of them, yes, but at that point I believe we challenge us to a better solution.

With your approach I think we currently have a problem with trailing whitespaces in unquoted values if I'm right, I will comment it in your PR.

I have an idea about a slight different approach, I will explain it after updating my PR.
And I will close some conversations above now, it gets a little confusing now. If you think some of them are still relevant you're welcome to reopen them.

Philipp-Binder commented:
> With your approach I think we currently have a problem with trailing whitespaces in unquoted values if I'm right, I will comment it in your PR.

Okay, I now understand why we do not have a problem with trailing whitespaces with your approach.
Of course the tests show that this is not the case, but understanding the Unquoted-Parser is quite hard now.
That's the only remaining problem then, and I will try to give a solution having that in mind.

One point where we currently differ:
You say my solution tries to check for inline comment and yours don't, but by checking for a `#` after some whitespaces of course you do check for inline comments. You just do not explicitly name it, but otherwise you wouldn't be able to prevent the trailing whitespaces.
Also you do check for linebreaks (aka controls) within Unquoted values, otherwise you wouldn't be able to stop the value before it. And this is kind of equal to quoted values, difference is onle that with quoted values is that we have the quotes as a defined start and endpoint instead of such constructs like `whitespaces(optional) followed by #, ' or "`

rogusdev commented:
On this one point: "check for" is different than "parse". Parsing *consumes* the characters, whereas checking for *prevents* consuming the characters. By not consuming comments in the unquoted parser, the Assignment parser can consume them instead, in a more consistent fashion with the other Value parsers.

Philipp-Binder commented:
Then we are on the same page, my approach using Except does not consume (aka Parse with your definition) comments too. That is still done by the Assignment parser.

```
InlineWhitespaceChars.Many().Then(w =>
                    NotControlNorWhitespaceChar("#'\"$").Once()
```
This piece of code of your approach parses InlineWhitespaces and if it parses any "disallowed" character afterwards the parser fails, resulting in throwing an exception or going on with an `Or(...)` or whatever is specified next, and parses frorm where the InlineWhitespace-Parser started, like a "rollback". The "not consume" is just because the parser fails, which is the same in my approach.

The real difference between the solutions is about the InlineWhitespace-Handling. and about the set of allowed characters inside an unquoted value.
Not sure if it is a must have, but the following test will fail with your solution.
```
Assert.Equal("allow 'whitespace' before 'quotation' inline", Parsers.UnquotedValue.End().Parse("allow 'whitespace' before 'quotation' inline").Value);
```
But except for this I support your approach as long as I do not find a better way to do it ;)

rogusdev commented:
Apparently there is a hot key to close a PR...? Not so helpful, github :(

rogusdev commented:
> Not sure if it is a must have, but the following test will fail with your solution.

I addressed this above on this PR here:
https://github.com/tonerdo/dotnet-env/pull/82#discussion_r1425612902

And I added an explicit comment to explain the tests I added to address this concern in my PR:
https://github.com/tonerdo/dotnet-env/pull/87/files#r1429207389

In short, I do not consider it a must have -- quite the opposite, it leads me to question some of the other things that are being allowed by these changes. All I wanted was for comments to not get consumed ;) I started expanding to these changes based on your interest, and because it actually made the code easier, but I consider it a slippery slope to allow quotes in unquoted values.

In truth, at this point, I am strongly leaning towards removing quote chars from unquoted vars because (in bash):
```
EVDNE=abc
echo $EVDNE
# abc
EVDNE=a b c
# b: command not found
EVDNE=a 'b' c
# b: command not found
```

Fish is a bit more forgiving:
```
set EVDNE abc
echo $EVDNE
# abc
set EVDNE a b c
echo $EVDNE
# a b c
set EVDNE a 'b' c
echo $EVDNE
# a b c
set EVDNE a'b'c
echo $EVDNE
# abc
```

But notice that even tho it does not complain about the `b` char like bash did for an unquoted value, it still does not include the quotes when that `'b'` is in quotes.

So, at this point, I am strongly leaning towards removing quotes from unquoted values, like it was before. They either break in bash env vars (which is what I want .env files to match) or else do not do the same thing in fish.

Thoughts?

rogusdev commented:
In fact, maybe I should just remove whitespace support from unquoted values entirely, given that example. We have quoted values to handle such things, and I want to be as close as possible to bash/posix/etc env var handling as possible.

rogusdev commented:
> I am strongly leaning towards removing quotes from unquoted values
and
> In fact, maybe I should just remove whitespace support from unquoted values entirely

I did both of these on my PR: https://github.com/tonerdo/dotnet-env/pull/87/commits/c21430928e52f062a6e12109631668f31be8e2ff

But then I reconsidered and re-added inline whitespace, to minimize breaking changes: https://github.com/tonerdo/dotnet-env/pull/87/commits/2a4fec859d437549b2e08a31d6cbf1b590f62fe0

But tbh I'm still not sure how I feel about inline whitespace. It does occur to me now that allowing this will break `source`-ing these files, but idk how important that is, and I also don't know how often people are using inline whitespace to begin with. So...

Right now I'm leaning towards merging my current PR, which allows inline whitespace, but prevents quote chars, in unquoted values, release that as 2.6.0 and then maybe release 3.0.0 as no inline whitespace. Thoughts?

Or, eh, just leave it all as-is, minimize the changes -- no one is complaining about inline whitespace, and no one is asking for more support for source-ing, so... I guess it ain't broke, and thus needs no fixing.

Philipp-Binder commented:
That's the ugly thing about the mising standard :D

My first intention was: yes, drop whitespace-support and take bash as de facto standard.

But my second thought was:
What is the real point for me to have an env-File.
It's mostly about having a super easy way to have env-vars at development-time without modifying the environment variables (and having problems with other projects and so on...)
And then I'm back to: let's have the user-side as easy as possible. And as far as I'm concerned, I use unquoted values nearly exclusively, just using quoted if theres no other way. I just want to paste my value and use it, not more, but not less.
And then I'm back at: let unquoted values have any value except for special leading chars (whitespace, quotes, hash) and no trailing quotes, that's it.

The thing is, should it be compatible with the capability of bash variable assignment?
A little bit of research gave me that Gist: https://gist.github.com/judy2k/7656bfe3b322d669ef75364a46327836
The discussion points at this repo, which tries to extend bash env-capabilities with using some definitions from other tools (like npm, docker etc, which all use their own definitions... quite ugly without a standard...): https://github.com/ko1nksm/shdotenv

That's also possibly not the best repo for a standard we want to use too, for example it does not allow whitespaces around the equal-sign, which is a good thing to be supported in my opinion. And it disallows whitespaces in unquoted values too (didn't try, but I expect it to be that way reading the readme).

What I try to say: I don't think that bash variable assignment is the best standard to be used, having local development/user experience in mind.

rogusdev commented:
Yeah, that's why I put whitespace back in. But I still think quotes in unquoted is a bad idea. Do you have an example case where you want to do that?

For me, I'm worried someone does this:
```
EVDNE=a'superimportant thing'
```

Where the leading `a` is a typo and would just be confusing for it to suddenly make the quotes somehow up in the value, rather than an error highlighting the typo. We have quoted values, we should encourage them, imo.

I am feeling very ok with my PR at this point. I am grateful for the discussion that has led here, we've covered good ground. But I want to 1) minimize sudden (breaking) changes and 2) keep the code clean and relatively straightforward to read. Fixing just the actual code cleanup goal, and supporting a couple of previously broken cases, but not changing behavior too much with allowing quote chars in unquoted suddenly, and certainly not removing whitespace support... I think this is a good change for a now venerable and otherwise very stable library.

Philipp-Binder committed:
rewrite UnquotedValue to disallow only certain leading characters; most other inline values are permitted (inline whitespaces/tab, quotes etc.)

Philipp-Binder commented:
I don't think a typo like that is a real good reason, because we are not supposed to cover typos I guess, and you will recognize this typo quite fast when you check the resulting value.

My example-case would be a passwort which is something like `D[}9P"LlBf!{Fg*%9S7]`.
You already have that annoying problem with Interpolation (which is a good thing, but really annoying with passwords).

But maybe the better strategy here is using singlequotes for passwords, because it additionally prevents unintended Interpolation.
With some highlighting (for example https://plugins.jetbrains.com/plugin/9525--env-files-support) you also see problems due to inline singlequotes quite good.
![grafik](https://github.com/tonerdo/dotnet-env/assets/52062680/f5c3c81a-3f3f-4fbf-a336-900c6d5349d5)

Saying that, I support you in disallowing quotes within unquoted values.

Unfortunately these .env-highlighting Plugins have their own definition due to the lack of standardization too... your example shows up like that:
![grafik](https://github.com/tonerdo/dotnet-env/assets/52062680/ac563dd8-0d1a-421a-ab78-1f3bdbefbd28)
The one with whitespaces complains only due to the inline whitespaces.
There are more ugly things with that analyzer, but that's not the point here of course ;)

Then there is only one thing I want to figure out now, and maybe preferring my PR due to that (after changing quote-behavior of course):
I would say it is a good thing to explicitly disallow certain leading-characters and add testcases for that, instead of relying on the behavior of the Assignment Parser. That is equivalent to DoubleQuoted and SingleQuoted which expect the " or ' as the first parsed character, whitespaces have to be handled before, otherwise the Values are not recognised. For UnquotedValues the difference is, that you do not have a special "initial" character but a set of explicitly disallowed ones.
I will comment that at your PR, including a test checking the behaviour.

Philipp-Binder committed:
disallow quotationChars inside UnquotedValue

Philipp-Binder committed:
include some tests, cleanup and readme from PR #87 and

rogusdev commented:
I look forward to your test case, because I am not sure what you mean about explicitly disallowing certain initial characters, other than what is already explicitly disallowed, and does have test cases for?

The rest all sounds good, and looks like the env extension agrees that my example is relevant, and a good reason to not allow quote chars.

Philipp-Binder commented:
Comment is added (https://github.com/tonerdo/dotnet-env/pull/87#issuecomment-1860547670).

I also updated my PR and doublechecked whether I have removed or changed tests.
==> there are just additional tests now, that can be discussed of course, but accordingly the solution is only slightly more strict compared to before, nothing additionally added.
As a bonus: all the examples are still working, so no removal of support for any given testcase.

Not sure if we removed anything now :D I forgot if quotes were somehow possible with the old solution, if so, that is the only breaking change, but that is intended as discussed.

So the main point of the PR is now the initial point: use Parser instead of Split/Trim and while at it, guard the UnquotedParser a little better.

Philipp-Binder committed:
added an additional test to show, that by interpolating you can get leading and trailing whitespaces within unquoted values now (before the Trim prevented that)

Philipp-Binder commented on test/DotNetEnv.Tests/EnvTests.cs:
```diff
@@ -177,7 +177,9 @@ public void ParseInterpolatedTest()
             Assert.Equal("test:testtest1\\uaeandtest1", Environment.GetEnvironmentVariable("TEST5_UNQUOTED"));
-
+            
+            Assert.Equal(" surrounded by spaces ", Environment.GetEnvironmentVariable("TEST_UNQUOTED_WITH_INTERPOLATED_SURROUNDING_SPACES"));
```
added this test to prove, that leading whitespaces are now present as we expect them to be when we interpolate within an unquoted value.
The Trimming was throwing them away before, so we get the expected output now compared to before.

rogusdev replied:
Oh that is good

rogusdev commented on src/DotNetEnv/Parsers.cs:
```diff
@@ -176,14 +177,16 @@ private static string ToUtf32Char (string hex)
-                .AtLeastOnce()
-                .Select(strs => new ValueActual(strs))
-            ).Many().Select(vs => new ValueCalculator(vs).Split("[ \t]#").Trim());
+            Parse.Regex("[ \t\"']").Not()
```
Why is this still here?

Philipp-Binder replied:
This is about the disallowed leading characters, and assures that we throw if there is a wrong character.
We can't do that inside, because due to the `Many` we would return an empty Value there.
And we need the `Many`, because it is fine to have no value, it is just not allowed to begin with one of these chars.

rogusdev replied:
Please walk me through why you think this implementation is better than mine for unquotedvalue.

Philipp-Binder replied:
In my opinion the main point is about the "completeness/correctness" of the unquotedValue (I try to reduce it to the core here, long version: https://github.com/tonerdo/dotnet-env/pull/87#issuecomment-1860793258), which results in the following:

1. User-Experience: if there is an unmatched quote, we get the "correct" exceptionmessage now, because unquotedValue will not succeed with "EmptyValue" and fails instead
2. There is a smaller risk of unexpected results if there is a bug within quotedValues, because unquotedValue is proven by Tests to fail if a value starts with a quote
3. So after all, an additional Test which prevents mistakes for any following PR

The main thing I'm not really happy with my own solution is about the `Then`.
If there is a way to make it more obvious why the parameterd in the `Then` are discarded that would be great...
But in general I like the explicit `Not`. I guess will try a solution like with QuotedValues, where the Contents are separated and a `From-Select` is used instead of `Then`.

Philipp-Binder replied:
```
        // unquoted values can have interpolated variables,
        // but only inline whitespace -- until a comment,
        // and no escaped chars, nor byte code chars
        private static readonly Parser<ValueCalculator> UnquotedValueContents =
            InterpolatedValue
                .Or(from inlineWhitespaces in InlineWhitespace
                    from _ in Parse.Char('#').Not()
                    from partOfValue in NotControlNorWhitespace("$\"'")
                    select new ValueActual(string.Concat(inlineWhitespaces, partOfValue)))
                .Many()
                .Select(vs => new ValueCalculator(vs));

        internal static readonly Parser<ValueCalculator> UnquotedValue =
            from _ in Parse.Regex("[ \t\"']").Not()
            from value in UnquotedValueContents
            select value;
```

This way we don't need to use the `Then`, which is better for readability I guess.
Due to the separation made by the `from` statements I would say we see the reason for discarding quite obvious now, equal to the QuotedValues, where we discard the quotes.

Just wanted to have you a look through it before pushing something new.
Tests still succeed, if I'm right we just replace `Then` by `from-select`.

rogusdev replied:
My personal interpretation of yours, in contrast to mine: https://github.com/tonerdo/dotnet-env/pull/87/files#r1437281239

Yours is (if I understand it correctly):
1. starts w a regex, but not-ed ("not" is a often confusing move in languages in general, and same in logic, imo) -- I'm not even sure exactly what this is supposed to match, obviously to prevent whitespace and quotes, but can't that be handled more directly by the matchers below? As in, why do you need this at all?
2. Then allow interpolated values -- I'm confused how this doesn't satisfy the initial regex+not to begin with, tbh
3. Or whitespace opening (same as me) followed by a char that is not a hash -- that then gets thrown away (`_2` is never concated) then it allows the same subset of chars mine does, with a similar concat to mine.

So, to my mind, yours has a confusing Regex at the start, which fills an unclear purpose, and then yours has a seemingly too wide `Char().Not()` that then gets thrown away too, and yet somehow seems to work regardless. To my mind, your approach is imply 1. less clear and 2. less consistent with how the other parsers work.

To your points, 1 would be very nice, tho I'm not sure I'm seeing that, and I would like you to point at the specific tests for 2 and 3 that you think are compelling that are not present in my PR -- nor even before either PR.

rogusdev replied:
OK, I added the regex to the start of mine ~https://github.com/tonerdo/dotnet-env/pull/87/commits/dc277bebea97f0378ab2f96eb05e3656db326dac~ https://github.com/tonerdo/dotnet-env/pull/87/commits/6cbcf89c61e194bfb983f0cc2d77c1770ec063d9 (now with just `Chars` instead of `Regex`) to understand what you were talking about, and I see now what you mean with the error messages being from unquoted values matching empty and then leaving unconsumed extra chars, vs quoted values matching but missing the final quote. That is an improvement, for sure. Thank you!

rogusdev replied:
I am still unclear what you mean by 2 and 3.

For 2, I did review my own new tests that fail with the previous code: https://github.com/tonerdo/dotnet-env/pull/87#issuecomment-1870680310 and that was already the case before, if I understand you correctly, and still is. Meaning, I am not clear what you are arguing about those tests are unique about your approach for `UnquotedValue`

For 3, I believe at this point we both have all the same tests (which we should) -- I only did not have your changes to the error messages, which I added once I understood what you meant and were aiming for.

rogusdev replied:
At this point, I like your UnquotedValue + UnquotedValueContents approach with from-select. I certainly want you to get the credit for the changes to the repo, as this is instigated by you :)

I do ask that you make sure you are including all the new tests I have (I believe we are all the same now, I just want you to double check). I do want you to use the exact error message tests I added to my PR tho, with the comment above, and the `abc` after the quote char to make the position of the failure more explicit -- and also the comment I added for the regex in my PR into your from-select approach as well, please.

rogusdev replied:
Oh, and to be clear: please replace `Parse.Regex("...").Not()` with simply `Parse.Chars(" \t\"'").Not()` as that remains cleaner for my tastes

Philipp-Binder replied:
First, thank you for giving me the credit for all of this stuff! It's quite some of your work and reviews too.

I will dive through all of it and work at the last todos.

Great that you found the reason behind the leading Not for the ErrorMessage on your own, I should have clarified that more precise and showing the exact test-case as you said. It's quite hard to find the correct "layer" of description if it's all so clear in the own head :D

About the Not(): In general I agree, that is definitely often confusing and that's the main reason why I was not really happy with the "Then" solution.
But with the "select-from" I think it is quite good readable and understandable now, it's just an initial check that we do not start with certain characters (yes. I will change it to Chars ;) ).
The reason why I would not try to move this checks inside Interpolated or the "default" ValueContent is about "separation of concerns" I would say. There is a single Statement now which says: "the beginning character is never allowed to be that". And if that is checked, we can go on with parsing the actual value.

It would have to be tested, but I think it is possible to to an `Except` at the end of the Parser instead of the `Not` in front of it, but I would prefer the solution with Not, because it avoids to think in the "wrong order" (it's a little bit like from inside to outside instead of fluent, but within fluent, so from behind to front there...)
The unused `_` and `_2` is just an empty object, because `Not` will never consume and return anything.

I guess I will post the `Except` solution afterwards for reference.

Philipp-Binder replied:
Now I see again why I would prefer my version with the Not for the value itself. I really focused on the outer one only before.

It's more or less equal to each other and I'm not sure which one to use now. Not sure if you like both from-selects or just the outer one ;)
I leave it with mine, including the comment for now waiting for your response.

Here is the reason:
With the `Not` we separate the check for the first character from the general character-check.
Means with my solution there is a single check for the `#` only, which doesn't consume anything. And then every character gets checked with `NotControlNorWhitespace("$\"'")`.
Your solution on the other hand checks and consumes the first character in a single step, so you must assure that the list of additionally excluded characters is "in sync" between both calls, except for the hash, which is relevant for the first character only.

Philipp-Binder replied:
For Quoted Values there is some separation of tests between Contents and Wrapper, should I do this too or should we stick with the current tests?

rogusdev replied:
I am fine with the Not() now that I understand that it is non consuming.

I don't feel a need for separating Content vs wrapper tests at this point. Let's just leave the tests we have now (again, making sure yours and mine are unified).

Philipp-Binder replied:
Now for reference, this is working too:
```
        internal static readonly Parser<ValueCalculator> UnquotedValue =
            UnquotedValueContents.Except(Parse.Chars(" \t\"'"));
```
Having the separation of UnquotedValue and UnquotedValueContent is is okay I would say, but I really dislike this order of statements, but it's kinda the same with Not I guess, which is put at the last position...
But with the separation from `select-from` this is quite explanative imho.

rogusdev commented on README.md:
````diff
@@ -225,6 +225,16 @@ env | grep TEST
 # TEST=value#notcomment
 ```
 
+However, unlike bash (but like fish), the `#` cannot be at the start, so as to not confuse with comments.
````
Thanks for merging my updated documentation, however, I intend to remove this whole section in my PR. I decided it is overly verbose.

Philipp-Binder replied:
Then I will update it again later ;)

Philipp-Binder replied:
I tried to make it less verbose, but I think it is worth to be mentioned.

Awaiting your response ;)

edit:
If you intend to remove the whole section (didn't read it like that at first) I can remove it entirely of course.
Maybe instead some examples about comments etc. are not too bad.
The usage section in this repo is quite good I think: https://github.com/ko1nksm/shdotenv#env-file-syntax
Has to be rewritten with examples matching for this solution, but it gives a really quick overview without having to read that much text...

Philipp-Binder replied:
I close this, if you are interested in changing the documentation completely at this point we should move it to an own Issue imho.

rogusdev commented on src/DotNetEnv/Parsers.cs:
```diff
@@ -176,14 +177,16 @@ private static string ToUtf32Char (string hex)
+                        .Or(InlineWhitespace.Then(inlineWhitespaces
+                            => Parse.Char('#').Not()
+                                .Then(_2 => NotControlNorWhitespace("$\"'"))
+                                .Select(valueString => (IValue)new ValueActual(string.Concat(inlineWhitespaces, valueString)))))
```
The cast should be unnecessary, and sprache has concat we should use at this point.

Philipp-Binder replied:
Yes, cast was remaining from before, I will remove it.

But I'm not sure what you mean regarding concat.
I had a look into your branch, and there you use concat from Linq, but that's because you work on char-Arrays and convert to a string by using `Text` afterwards.
Here we have strings, not char-Arrays, and we expect a string as response.
I can use concat on the strings of course, but that will return a char-Array, which I need to convert to a string afterwards.

If there is something different I just don't see, please give me some hint, or a code-suggestion.

Philipp-Binder replied:
Updated, leaving it open for your response about concat

rogusdev replied:
Ah, indeed, I was uncertain what types you had at this point. I'm fine resolving this thread, as at this point I'm more interested in your answer to: https://github.com/tonerdo/dotnet-env/pull/82#discussion_r1432311810

Philipp-Binder committed:
refactor according to review
- update README.md (less verbose, but still mentioning)
- remove unnecessary cast

rogusdev commented on README.md:
```diff
@@ -225,7 +225,11 @@ env | grep TEST
+This is because whitespaces between EqualSign and Value are allowed, which is not allowed in bash.
+This prevents confusion between `KEY=#comment` and `KEY= #comment`, which is expected to give the same result when leading whitespaces before the value are allowed.
+
+And, unlike bash, inline whitespace is allowed so you can do:
```
`also unlike` for flow, please. Otherwise, I like the changes

rogusdev replied:
I updated the readme copy on my PR, incorporating your framing -- but still with changes, I'd like for you to copy over, please.

Philipp-Binder committed:
replace usage of `Then` with `select-from`

rogusdev commented on test/DotNetEnv.Tests/ParserTests.cs:
```diff
@@ -241,6 +241,23 @@ public void ParseUnquotedValue ()
+            Assert.Equal("", Parsers.UnquotedValue.Parse("#commentOnly").Value);
+
+            // prevent quotationChars inside unquoted values
+            Assert.Throws<ParseException>(() => Parsers.Assignment.End().Parse("EV_DNE=a'b'c"));
```
These tests are in the wrong function -- this is `ParseUnquotedValue` and should only have tests for `Parsers.UnquotedValue` -- not anything for `Parses.Assignment` as you added here

rogusdev commented on test/DotNetEnv.Tests/ParserTests.cs:
```diff
@@ -361,6 +386,21 @@ public void ParseAssignment ()
             Assert.Throws<ParseException>(() => Parsers.Assignment.End().Parse("EV_DNE='"));
             Assert.Throws<ParseException>(() => Parsers.Assignment.End().Parse("EV_DNE=0\n1"));
 
+            testParse("EV_DNE", "a'b'' 'c' d", "EV_DNE=\"a'b'' 'c' d\" #allow inline singleQuotes in doubleQuoted values");
```
I don't want to call these "inline" when they are inside a (single/double) quoted value. So, aside from the terminology, I think these are good additions.

Philipp-Binder committed:
refactor according to review

Philipp-Binder commented:
Red-Green Test-List:

It's more or less the same list as in the "twin"-PR (compare https://github.com/tonerdo/dotnet-env/pull/87#issuecomment-1870680310)

- EnvTests
  - ParseInterpolatedTest
    - `Assert.Equal(" surrounded by spaces ", Environment.GetEnvironmentVariable("TEST_UNQUOTED_WITH_INTERPOLATED_SURROUNDING_SPACES"));` --> old version trimmed the final calculated value
    - `Assert.Equal("Parsing failure: Unexpected end of input reached; expected ' (Line 1, Column 20); recently consumed: QUOTE='abc", ex.Message);` --> message changed (better user-experience)
    - `Assert.Equal("Parsing failure: Unexpected end of input reached; expected \" (Line 1, Column 20); recently consumed: QUOTE=\"abc", ex.Message);` --> message changed (better user-experience)
- ParserTests
  - ParseUnquotedValue
    - `Assert.Throws<ParseException>(() => Parsers.UnquotedValue.Parse("'"));` --> old version returned empty value
    - `Assert.Throws<ParseException>(() => Parsers.UnquotedValue.Parse("\""));` --> old version returned empty value
    - `Assert.Throws<ParseException>(() => Parsers.UnquotedValue.Parse(" "));` --> old version returned empty value
    - `Assert.Throws<ParseException>(() => Parsers.UnquotedValue.Parse("\t"));` --> old version returned empty value
    - `Assert.Equal("", Parsers.UnquotedValue.Parse("#").Value); // no value, empty comment` --> old version treated this as value
    - `Assert.Equal("", Parsers.UnquotedValue.Parse("#commentOnly").Value);` --> old version treated this as value
  - ParseValue
    - `Assert.Throws<ParseException>(() => Parsers.Value.Parse("'"));` --> old version returned empty value
    - `Assert.Throws<ParseException>(() => Parsers.Value.Parse("\""));` --> old version returned empty value
  - ParseAssignment
    - `testParse("EV_DNE", "", "EV_DNE=#no value just comment");` --> old version treated this as value
    - `testParse("EV_DNE", "", "EV_DNE= #no value just comment");` --> old version treated this as value
    - `testParse("EV_DNE", "test", "EV_DNE= test  #a'bc allow singleQuotes in comment");` --> old version did not allow this
    - `testParse("EV_DNE", "test", "EV_DNE= test  #a\"bc allow doubleQuotes in comment");` --> old version did not allow this

Philipp-Binder commented:
I tried to catch all todos, hopefully I did not forget something, otherwise please mention it!

Last open Thread I guess is `is UnquotedValue/UnquotedValueContent okay now` --> https://github.com/tonerdo/dotnet-env/pull/82#discussion_r1437756522.

edit: did not see that you already answered ;)
Then I guess I await a final review (given I did not forget something...) 👍

rogusdev commented on src/DotNetEnv/Parsers.cs:
```diff
@@ -173,17 +174,25 @@ private static string ToUtf32Char (string hex)
-                .AtLeastOnce()
-                .Select(strs => new ValueActual(strs))
-            ).Many().Select(vs => new ValueCalculator(vs).Split("[ \t]#").Trim());
+            from _ in Parse.Chars(" \t\"'").Not()
```
Please include my comments here, from: https://github.com/tonerdo/dotnet-env/pull/87/files#r1438698673

rogusdev commented on src/DotNetEnv/Parsers.cs:
```diff
@@ -173,17 +174,25 @@ private static string ToUtf32Char (string hex)
+            InterpolatedValue
+                .Or(from inlineWhitespaces in InlineWhitespace
+                    from _ in Parse.Char('#').Not() // "#" after a whitespace is the beginning of a comment --> not allowed
+                    from partOfValue in NotControlNorWhitespace("$\"'") // quotes are not allowed in values, because in a shell they lead to inconsistent results
```
They are not "inconsistent results" they are "different results than what they would mean here". Please use language closer to mine

Philipp-Binder replied:
Updated, should be okay now I guess.

rogusdev commented on test/DotNetEnv.Tests/ParserTests.cs:
```diff
@@ -312,6 +331,12 @@ public void ParseValue ()
+            
+            // throw on unmatched quotes
+            Assert.Throws<ParseException>(() => Parsers.Value.Parse("'"));
+            Assert.Throws<ParseException>(() => Parsers.Value.Parse("\""));
```
I believe I added a few more here -- were they duplicates? Or do you think they are not needed?

Philipp-Binder replied:
Comments at the discussion below, closing this

rogusdev commented on test/DotNetEnv.Tests/ParserTests.cs:
```diff
@@ -361,6 +387,21 @@ public void ParseAssignment ()
+            testParse("EV_DNE", "test", "EV_DNE= test #a$bc allow dollarSign in comment");
+
+            testParse("EV_DNE", "http://www.google.com/#anchor", "EV_DNE=http://www.google.com/#anchor #inline hash is part of value");
+
```
Please just double check that we both added all the same tests, in more or less all the same places.

Philipp-Binder replied:
- ParseUnquotedValue
  - reordered and at some tests removed the "End", because it is not needed (and we expect the same result in both cases, but with the End we cover only the End, while without we expect it to throw if it reads until the end or not, it doesn't matter)
- ParseValue
  - the "whitespace throws"-Test was missing, added it; (without "End" again because it wasn't necessary to have it throwing)
  - remove the "duplicated" unmatched quote-test with the "End"-statement. That is obsolete because it doesn't test additional actual behaviour.
- ParseAssignment
  - added `testParse("EV_DNE", "a#b#c", "EV_DNE=a#b#c # comment");` (changed the comment to be more explaining)
  - renamed "inline hash" to "inner hash", according to your comment about "inline quotes", which is also not real inline, but inner
  - there are 4 tests with a comment "already working"; they are duplicates compared to these two tests:
    - `testParse("EV_DNE", "a'b'' 'c' d", "EV_DNE=\"a'b'' 'c' d\" #allow singleQuotes in doubleQuoted values");`
    - `testParse("EV_DNE", "a\"b\"\" \"c\" d", "EV_DNE='a\"b\"\" \"c\" d' #allow doubleQuotes in singleQuoted values");`
  - added `testParse("EV_DNE", "a\"b\"\" \"c\" d", "EV_DNE=\"a\\\"b\\\"\\\" \\\"c\\\" d\" #allow quoted doubleQuotes in doubleQuoted values");` (quite good test, did not see this one before; fixed the comment tho, should be "escaped doubleQuotes" instead of "quoted doubleQuotes")
  - added `Assert.Throws<ParseException>(() => Parsers.Assignment.End().Parse("EV_DNE='a'b'' 'c' d'"));  // no single quotes inside single quoted values`
    - added an equivalent test for (unescaped) doubleQuotes inside doubleQuoted too
    - removed the "End", not necessary to let it throw here too ;)
  - added `testParse("EV_DNE", "a#b#c# not a comment", "EV_DNE=a#b#c# not a comment");`
  - I did not add the the huge comment and the corresponding tests for inner quotes in unquoted values; imho this is well tested in `ParseUnquotedValue` directly; that's also true for quite some other things, many stuff in assignment is duplicated from other methods; but due to the huge comment I decided to not adding this --> open for other thoughts ;)

That should be all, pushing it in a few seconds, wrote that while changing the code accordingly ;)

rogusdev replied:
Sounds good, thanks!

Philipp-Binder committed:
add additional tests and update comments according to review

Philipp-Binder committed:
reordered some tests to allow better comparing vs PR #87
